File: LeetCode/Python/Medium/513-Find-Bottom-Left-Tree-Value/solution.py
from collections import deque

class Solution:
    def findBottomLeftValue(self, root: Optional[TreeNode]) -> int:
        result=0

        last_level=self.get_height(root)

        result=self.get_leftmost_node_in_bottom_row(root, last_level)

        return result

    def get_height(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0

        left_subtree_height=self.get_height(root.left)

        right_subtree_height=self.get_height(root.right)

        return 1 + max(left_subtree_height, right_subtree_height)

    def get_leftmost_node_in_bottom_row(self, root: Optional[TreeNode], last_level: int) -> int:
        queue=deque()

        queue.append(root)

        level=1

        left_most_node=-1

        while queue:
            q_size=len(queue)

            for iteration in range(0, q_size):
                node=queue.popleft()

                if level==last_level and iteration==0:
                    left_most_node=node.val

                    break

                if node.left:
                    queue.append(node.left)

                if node.right:
                    queue.append(node.right)

            if queue:
                level+=1

        return left_most_node

File: LeetCode/Rust/Medium/513-Find-Bottom-Left-Tree-Value/solution.rs
use std::collections::{VecDeque};
use std::cmp;
use std::rc::Rc;
use std::cell::{Ref, RefCell};
impl Solution
{
    pub fn find_bottom_left_value(root: Option<Rc<RefCell<TreeNode>>>) -> i32
    {
        let mut result=-1;

        let height: i32=Self::get_height(&root);

        result=Self::get_leftmost_node_in_bottom_row(&root, height);

        result
    }

    fn get_height(root: &Option<Rc<RefCell<TreeNode>>>) -> i32
    {
        if root.is_none()
        {
            return 0;
        }

        let root_reference_counter: &Rc<RefCell<TreeNode>>=root.as_ref().unwrap();

        let root_ref: Ref<'_, TreeNode>=root_reference_counter.borrow();

        let left_subtree_height: i32=Self::get_height(&root_ref.left);

        let right_subtree_height: i32=Self::get_height(&root_ref.right);

        1 + std::cmp::max(left_subtree_height, right_subtree_height)
    }

    fn get_leftmost_node_in_bottom_row(root: &Option<Rc<RefCell<TreeNode>>>, last_level: i32) -> i32
    {
        let mut leftmost_node: i32=-1;

        let mut queue: VecDeque<Rc<RefCell<TreeNode>>>=VecDeque::new();

        queue.push_back(Rc::clone(root.as_ref().unwrap()));

        let mut level: i32=1;

        while !queue.is_empty()
        {
            let q_size: usize=queue.len();

            for iteration in 0..q_size
            {
                if let Some(node)=queue.pop_front()
                {
                    let node_ref: Ref<'_, TreeNode>=node.borrow();

                    if iteration==0 && level==last_level
                    {
                        leftmost_node=node_ref.val;

                        break;
                    }

                    if node_ref.left.is_some()
                    {
                        queue.push_back(Rc::clone(node_ref.left.as_ref().unwrap()));
                    }

                    if node_ref.right.is_some()
                    {
                        queue.push_back(Rc::clone(node_ref.right.as_ref().unwrap()));
                    }
                }
            }

            if !queue.is_empty()
            {
                level+=1;
            }
        }

        leftmost_node
    }
}